core/container/externalbuilder/instance.go

```diff
@@ -186,7 +186,13 @@ func (i *Instance) Wait() (int, error) {
 		return -1, errors.Errorf("instance was not successfully started")
 	}
 
+	i.Session.WaitGroup.Add(1)
 	err := i.Session.Wait()
+	i.Session.WaitGroup.Done()
+	// have the Instance wait for all goroutines waiting on the Session
+	i.Session.WaitGroup.Wait()
+	// signal that this chaincode instance is done (i.e. successfully cleaned up)
+	i.Session.doneChan <- struct{}{}
```

Comment:
ironically, your comment is exactly what `Wait` does. It's the implementation that's making things more confusing.